catalyst/contrib/nn/criterion/ntxent.py

```diff
@@ -0,0 +1,74 @@
+import torch
+from torch import nn
+
+
+class NTXentLoss(nn.Module):
+    """A Contrastive embedding loss.
+
+    It has been proposed in `A Simple Framework
+    for Contrastive Learning of Visual Representations`_.
+
+    Example:
+
+    .. code-block:: python
+
+        import torch
+        from torch.nn import functional as F
+        from catalyst.contrib.nn import NTXentLoss
+
+        embeddings_left = F.normalize(torch.rand(256, 64, requires_grad=True))
+        embeddings_right = F.normalize(torch.rand(256, 64, requires_grad=True))
+        criterion = NTXentLoss(tau = 0.1)
+        criterion(embeddings_left, embeddings_right)
+
+    .. _`A Simple Framework for Contrastive Learning of Visual Representations`:
+        https://arxiv.org/abs/2103.03230
+    """
+
+    def __init__(self, tau: float, reduction: str = "mean") -> None:
+        """
+
+        Args:
+            tau: temperature
+            reduction (string, optional): specifies the reduction to apply to the output:
+                ``"none"`` | ``"mean"`` | ``"sum"``.
+                ``"none"``: no reduction will be applied,
+                ``"mean"``: the sum of the output will be divided by the number of
+                positive pairs in the output,
+                ``"sum"``: the output will be summed.
+        """
+        super().__init__()
+        self.tau = tau
+        self.cosine_sim = nn.CosineSimilarity()
+        self.reduction = reduction
+
+    def forward(self, features1: torch.Tensor, features2: torch.Tensor) -> torch.Tensor:
```

Comment:
Okay, I am ready to write my concerns more formally.
Please, check my formulas.

Some notation:
f1 - features1
f2 - features2
<> - cosine similarity (without exp function)
N - batch size
![] - Iverson notation
\tau = 1, for simplicity

According to the paper, the final loss is:
```
L = 1/(2N) \sum_{k=0}^{N-1} ( l(f1[k], f2[k]) + l(f2[k], f1[k]) )
```

where
```

l(f[i], g[i]) = -<f[i], g[i]> + log( \sum_{k=0}^{N-1} ( ![k \neq i] * exp( <f[i], f[k]> ))) + log( \sum_{k=0}^{N-1}( exp( ( <f[i], g[k]>))))

```

Now, this function returns loss, which is:

```
L = 1/N \sum_{k=0}^{N-1} ( l(f1[k], f2[k]) )
```

and `l(f[i], g[i])` is missing `log( \sum_{k=0}^{N-1} ( ![k \neq i] * exp( <f[i], f[k]> )))` term

Reply:
I think I have fixed the loss issues.
Could you check the fix?

Reply:
okay, now it seems great for me, except I don't quite understand the cosine matrix building. Can you explain it to me, please?

Reply:
I have decided to do it that way because I haven't found the native pair-wise cosine matrix in PyTorch.
So I decided to use torch.cdist (the idea is pretty simple):
if ||x|| = ||y|| = 1 then we have ||x-y||^2 = <x-y, x-y> = 2 - 2<x,y>.

Reply:
Oh, now I get it. It's really cool idea. But I think my proposal (with torch.mm) is a little bit simpler. It's totally up to you, what you decide to choose (but I suggest you leave some comments, with explanation, if you decide to go with torch.cdist)